src/worker/plugins/loadPlugin.ts

```diff
@@ -94,3 +102,51 @@ export async function loadPlugin(server: PluginsServer, pluginConfig: PluginConf
     }
     return false
 }
+
+async function inferPluginCapabilities(server: PluginsServer, pluginConfig: PluginConfig, prevConfig?: PluginConfig) {
+    // infer on load implies there's no lazy loading, but all workers get
+    // these properties loaded
+    const vm = await pluginConfig.vm?.resolveInternalVm
+    const capabilities: Capabilities = { scheduled_tasks: [], jobs: [], methods: [] }
+
+    const tasks = vm?.tasks
+    const methods = vm?.methods
+
+    if (methods) {
+        for (const [key, value] of Object.entries(methods)) {
+            if (value !== undefined) {
+                capabilities.methods!.push(key)
+            }
+        }
+    }
+
+    if (tasks?.schedule) {
+        for (const [key, value] of Object.entries(tasks.schedule)) {
+            if (value) {
+                capabilities.scheduled_tasks!.push(key)
+            }
+        }
+    }
+
+    if (tasks?.job) {
+        for (const [key, value] of Object.entries(tasks.job)) {
+            if (value) {
+                capabilities.jobs!.push(key)
+            }
+        }
+    }
+
+    const prevCapabilities = prevConfig?.plugin?.capabilities
+    if (
+        prevCapabilities &&
+        capabilities.jobs?.sort().toString() == prevCapabilities.jobs?.sort().toString() &&
+        capabilities.scheduled_tasks?.sort().toString() == prevCapabilities.scheduled_tasks?.sort().toString() &&
+        capabilities.methods?.sort().toString() == prevCapabilities.methods?.sort().toString()
+    ) {
+        // pass - no change in capabilities
+    } else {
+        await setPluginCapabilities(server, pluginConfig, capabilities)
+    }
+
+    pluginConfig.plugin!.capabilities = capabilities
```

Comment:
Here again the `!` could cause subtle issues. After all, we're not sure this function will always be called with pluginConfigs that contain a plugin. It might be the case when we follow the current trail that leads to this function, but it might not be the case in the future.

Often the solution is to use a guard, such as `if (!pluginConfig.plugin) { throw new Error('how the hell did you get here?') }`, or just `return`, depends on the case. Guards are best placed at the top of a function as well.